Data/Graph/Inductive/Query/Cycles.hs

```diff
@@ -0,0 +1,280 @@
+{- |
+   Module      : Data.Graph.Inductive.Query.Cycles
+   Description : Algorithms for finding all cycles.
+   Copyright   : (c) Gabriel Hjort Blindell 2017
+   License     : 2-Clause BSD
+   Maintainer  : gabriel.hjort.blindell@gmail.com
+
+   Defines algorithms that find all cycles in a given graph.
+ -}
+module Data.Graph.Inductive.Query.Cycles
+  ( cyclesIn
+  , cyclesIn'
+  , strongComponentsOf
+  , uniqueCycles
+  , uniqueCycles'
+  )
+where
+
+import Data.Graph.Inductive.Graph
+
+import Data.List
+  ( (\\)
+  , delete
+  , tails
+  )
+import Data.Maybe
+  ( fromJust )
+import Control.Monad
+  ( ap )
+import qualified Data.Map as M
+
+
+
+-- | Obtain the labels for a list of 'Node's.
+--   It is assumed that each 'Node' is indeed present in the given graph.
+addLabels :: (Graph g) => g a b -> [Node] -> [LNode a]
+addLabels gr = map (ap (,) (fromJust . lab gr))
+
+twoCycle      :: (Graph g) => g a b -> Node -> [Node]
+twoCycle gr n = filter (elem n . suc gr) (delete n $ suc gr n)
+
+-- | Determines if the list of nodes represents a regular subgraph.
+isRegular :: (Graph g) => g a b -> [Node] -> Bool
+isRegular g ns = all allTwoCycle split
+    where
+      -- Node + Rest of list
+      split = zip ns tns'
+      tns' = tail $ tails ns
+      allTwoCycle (n,rs) = null $ rs \\ twoCycle g n
+
+-- | Contains the necessary data structures used by 'strongComponentsOf'.
+data SCCState g a b
+  = SCCState
+      { sccComponents :: [g a b]
+        -- ^ The components found so far.
+      , sccCurrentIndex :: Int
+        -- ^ The current index.
+      , sccStack :: [Node]
+        -- ^ The node stack.
+      , sccNodeInfo :: M.Map Node (Bool, Int, Int)
```

Comment:
I think a specific data structure with record labels would work better than a triple here; it would not only provide documentation as to what each field means, but you can then use them as selectors rather than pattern matching on the entire triple to get values out.